themes/default/content/docs/languages-sdks/python/python-blocking-async.md

```diff
@@ -0,0 +1,199 @@
+from being made by the Pulumi runtime by preventing the event loop from
+processing asynchronous tasks.  
+
+If the output from the blocking code will be used as an input to a resource,
+then the blocking code may be executed on another thread with the future result
+captured as a `pulumi.Output` object. In the same way, asynchronous code (using
+Python’s [`asyncio`](https://docs.python.org/3/library/asyncio.html)) can also
+be evaluated concurrently with the Pulumi runtime and the future result also
+captured using `pulumi.Output`.
+
+## Background
+
+As covered in [Concepts](/docs/intro/concepts), when a Pulumi program is run it creates resources and the dependencies or
+connections between them. Consider the following example (taken from the Pulumi
+blog article [Programming the Cloud with Python](/blog/programming-the-cloud-with-python))
+
+```python
+import mimetypes
+import os
+
+from pulumi import export, FileAsset
+from pulumi_aws import s3
+
+web_bucket = s3.Bucket('s3-website-bucket', website={
+    "index_document": "index.html"
+})
+
+content_dir = "www"
+for file in os.listdir(content_dir):
+    filepath = os.path.join(content_dir, file)
+    mime_type, _ = mimetypes.guess_type(filepath)
+    obj = s3.BucketObject(file,
+        bucket=web_bucket.id,
+        source=FileAsset(filepath),
+        content_type=mime_type)
+```
+
+Here the resources are an S3 Bucket and some number of S3 BucketObjects---one
+for each file in the `www` directory. The bucket objects are told which bucket
+they will be put in using the id of the bucket. This establishes the dependency
+between the BucketObject and the Bucket.
+
+The type of the bucket id is `Output[str]`. The
+[`pulumi.Output`](/docs/reference/pkg/python/pulumi/#pulumi.Output) type holds a
+piece of data, in this case the id of the bucket, this is not necessarily known
+during the execution of the program, and is held internally within the output
+object as a future---the awaitable result of an asynchronous operation.
+
+In this way the Pulumi program can proceed and build the graph of resources
+without having to wait for each resource in turn. When Pulumi is previewing a
+stack the value of an output may never be known. When resources are then created
+that can be done concurrently according to the dependency graph that the Pulumi
+program defines.
+
+## Blocking and Asynchronous Code
+
+The example above, and many or most Python Pulumi programs, contains no explicit
+asynchronous code. While the resources are connected using futures, the
+asynchronous evaluation all happens behind the scenes.
+
+This can cause problems both with blocking code and with explicitly asynchronous code
+
+- Blocking code will prevent the event loop on the Pulumi program’s thread from
+  pumping while the blocking code is executing
+- Asynchronous code may not be executed unless it is explicitly scheduled. It is
+  not appropriate to call `asyncio.run` from within a Pulumi program
```

Comment:
What happens?

Reply:
It will raise an exception because there is already an event loop

Reply:
Updated the wording here.